File: setup/github-app.ts
/**
 * GitHub App Setup (CLI)
 *
 * Creates a GitHub App via the manifest flow without exposing the setup
 * on the main webhook server. Runs a temporary local HTTP server to
 * serve the manifest form and receive the OAuth callback, then writes
 * credentials to .env and exits.
 *
 * Usage: npx tsx setup/index.ts --step github-app [-- --webhook-url URL]
 */
import { execSync } from 'child_process';
import fs from 'fs';
import http from 'http';
import os from 'os';
import path from 'path';

import { buildAppManifest } from '../src/github/app-manifest.js';
import { logger } from '../src/logger.js';
import { readEnvFile } from '../src/env.js';
import { emitStatus } from './status.js';

const SETUP_PORT = 23847; // Arbitrary high port unlikely to conflict

function openBrowser(url: string): void {
  const platform = os.platform();
  try {
    if (platform === 'darwin') {
      execSync(`open ${JSON.stringify(url)}`);
    } else if (platform === 'win32') {
      execSync(`start "" ${JSON.stringify(url)}`);
    } else {
      execSync(`xdg-open ${JSON.stringify(url)} 2>/dev/null || sensible-browser ${JSON.stringify(url)} 2>/dev/null`);
    }
  } catch {
    // Browser open failed — user will need to navigate manually
  }
}

async function exchangeCode(code: string): Promise<{
  id: number;
  slug: string;
  pem: string;
  webhook_secret: string;
  html_url: string;
}> {
  const response = await fetch(`https://api.github.com/app-manifests/${code}/conversions`, {
    method: 'POST',
    headers: { Accept: 'application/vnd.github+json' },
  });

  if (!response.ok) {
    const body = await response.text();
    throw new Error(`GitHub API error: ${response.status} ${body}`);
  }

  return response.json() as Promise<{
    id: number;
    slug: string;
    pem: string;
    webhook_secret: string;
    html_url: string;
  }>;
}

function saveCredentials(data: {
  id: number;
  slug: string;
  pem: string;
  webhook_secret: string;
}): void {
  const configDir = path.join(os.homedir(), '.config', 'codeclaw');
  fs.mkdirSync(configDir, { recursive: true });
  const pemPath = path.join(configDir, 'github-app.pem');
  fs.writeFileSync(pemPath, data.pem, { mode: 0o600 });
  logger.info({ pemPath }, 'GitHub App private key saved');

  const envPath = path.join(process.cwd(), '.env');
  const envLines = [
    '',
    '# GitHub App (auto-configured)',
    `GITHUB_APP_ID=${data.id}`,
    `GITHUB_WEBHOOK_SECRET=${data.webhook_secret}`,
    `GITHUB_PRIVATE_KEY_PATH=${pemPath}`,
    '',
  ].join('\n');

  fs.appendFileSync(envPath, envLines);
  logger.info({ appId: data.id, slug: data.slug }, 'GitHub App credentials saved to .env');
}

function buildSetupPageHtml(webhookUrl: string): string {
  const callbackUrl = `http://localhost:${SETUP_PORT}/callback`;
  const manifest = buildAppManifest(webhookUrl);
  // Override redirect_url to point to our local setup server
  (manifest as Record<string, unknown>).redirect_url = callbackUrl;
  const manifestJson = JSON.stringify(manifest);

  return `<!DOCTYPE html>
<html>
<head><title>CodeClaw &mdash; Create GitHub App</title>
<style>
  body { font-family: system-ui; max-width: 600px; margin: 40px auto; padding: 0 20px; }
  h1 { color: #333; }
  .btn { background: #2ea44f; color: white; border: none; padding: 12px 24px;
         font-size: 16px; border-radius: 6px; cursor: pointer; }
  .btn:hover { background: #2c974b; }
  code { background: #f6f8fa; padding: 2px 6px; border-radius: 3px; }
  .note { background: #fff8c5; border: 1px solid #d4a72c; border-radius: 6px;
          padding: 12px 16px; margin: 16px 0; font-size: 14px; }
</style>
</head>
<body>
  <h1>CodeClaw &mdash; Create GitHub App</h1>
  <p>Click below to create a GitHub App with the correct permissions.</p>
  <p>Webhook URL: <code>${webhookUrl}/github/webhooks</code></p>
  <form action="https://github.com/settings/apps/new" method="post">
    <input type="hidden" name="manifest" value='${manifestJson.replace(/'/g, '&#39;')}'>
    <button type="submit" class="btn">Create GitHub App</button>
  </form>
  <p><small>This will redirect you to GitHub to approve the app creation.</small></p>
</body>
</html>`;
}

function buildSuccessHtml(slug: string, installUrl: string): string {
  return `<!DOCTYPE html>
<html>
<head><title>CodeClaw &mdash; Setup Complete</title>
<style>
  body { font-family: system-ui; max-width: 600px; margin: 40px auto; padding: 0 20px; }
  h1 { color: #2ea44f; }
  .btn { background: #2ea44f; color: white; border: none; padding: 12px 24px;
         font-size: 16px; border-radius: 6px; cursor: pointer; text-decoration: none;
         display: inline-block; }
  code { background: #f6f8fa; padding: 2px 6px; border-radius: 3px; }
</style>
</head>
<body>
  <h1>Setup Complete!</h1>
  <p>GitHub App <strong>${slug}</strong> has been created. Credentials saved to <code>.env</code>.</p>
  <p>Now install it on the repositories you want the bot to monitor:</p>
  <a href="${installUrl}" class="btn">Install on Repositories</a>
  <p><small>You can close this tab after installing. Then restart CodeClaw to load the new credentials.</small></p>
</body>
</html>`;
}

export async function run(args: string[]): Promise<void> {
  // Check if already configured
  const env = readEnvFile(['GITHUB_APP_ID']);
  if (env.GITHUB_APP_ID) {
    emitStatus('GITHUB_APP', {
      STATUS: 'already_configured',
      APP_ID: env.GITHUB_APP_ID,
    });
    return;
  }

  // Get webhook URL from args or fail
  const urlIdx = args.indexOf('--webhook-url');
  const webhookUrl = urlIdx !== -1 ? args[urlIdx + 1] : undefined;

  if (!webhookUrl) {
    emitStatus('GITHUB_APP', {
      STATUS: 'failed',
      ERROR: 'Missing --webhook-url. Provide the public URL where GitHub will send webhooks (e.g. https://abc123.ngrok-free.app).',
    });
    process.exit(1);
  }

  // Validate URL
  try {
    new URL(webhookUrl);
  } catch {
    emitStatus('GITHUB_APP', {
      STATUS: 'failed',
      ERROR: `Invalid URL: ${webhookUrl}`,
    });
    process.exit(1);
  }

  // Start temporary local server
  const html = buildSetupPageHtml(webhookUrl);

  await new Promise<void>((resolve, reject) => {
    const server = http.createServer(async (req, res) => {
      const url = new URL(req.url || '/', `http://localhost:${SETUP_PORT}`);

      if (url.pathname === '/' || url.pathname === '/setup') {
        res.writeHead(200, { 'Content-Type': 'text/html' });
        res.end(html);
        return;
      }

      if (url.pathname === '/callback') {
        const code = url.searchParams.get('code');
        if (!code) {
          res.writeHead(400, { 'Content-Type': 'text/plain' });
          res.end('Missing code parameter');
          return;
        }

        try {
          const data = await exchangeCode(code);
          saveCredentials(data);

          const installUrl = `${data.html_url}/installations/new`;
          res.writeHead(200, { 'Content-Type': 'text/html' });
          res.end(buildSuccessHtml(data.slug, installUrl));

          emitStatus('GITHUB_APP', {
            STATUS: 'ok',
            APP_ID: data.id,
            SLUG: data.slug,
            INSTALL_URL: installUrl,
          });

          // Give the browser a moment to receive the response, then shut down
          setTimeout(() => {
            server.close();
            resolve();
          }, 1000);
        } catch (err) {
          const message = err instanceof Error ? err.message : String(err);
          logger.error({ err }, 'GitHub App manifest callback failed');
          res.writeHead(500, { 'Content-Type': 'text/plain' });
          res.end(`Setup failed: ${message}`);
          server.close();
          reject(err);
        }
        return;
      }

      res.writeHead(404, { 'Content-Type': 'text/plain' });
      res.end('Not found');
    });

    server.listen(SETUP_PORT, '127.0.0.1', () => {
      const setupUrl = `http://localhost:${SETUP_PORT}/setup`;
      console.log(`\nOpening browser to create GitHub App...\n  ${setupUrl}\n`);
      openBrowser(setupUrl);
    });

    // Timeout after 5 minutes
    setTimeout(() => {
      server.close();
      reject(new Error('Setup timed out — no callback received within 5 minutes.'));
    }, 5 * 60 * 1000);
  });
}

File: setup/github_app.py
"""GitHub App Setup (CLI).

Creates a GitHub App via the manifest flow. Runs a temporary local HTTP
server to serve the manifest form and receive the OAuth callback, then
writes credentials to .env and exits.

Usage: python -m setup.github_app --webhook-url <URL>
"""

from __future__ import annotations

import argparse
import json
import sys
import threading
import webbrowser
from functools import partial
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

import httpx

from clawcode.env import read_env_file
from clawcode.logger import logger

SETUP_PORT = 23847  # Arbitrary high port unlikely to conflict
TIMEOUT_SECONDS = 5 * 60  # 5 minutes


# ---------------------------------------------------------------------------
# Manifest
# ---------------------------------------------------------------------------

def _is_public_url(url: str) -> bool:
    try:
        hostname = urlparse(url).hostname
        return hostname not in ("localhost", "127.0.0.1", "::1")
    except Exception:
        return False


def build_app_manifest(webhook_url: str) -> dict:
    """Build the GitHub App manifest for the one-click creation flow.

    When the webhook URL is localhost we omit hook_attributes — the setup
    page prompts the user for a public tunnel URL instead.
    """
    callback_url = f"http://localhost:{SETUP_PORT}/callback"
    manifest: dict = {
        "name": "ClawCode AI",
        "url": "https://github.com/nadahlberg/clawcode",
        "redirect_url": callback_url,
        "public": False,
        "default_permissions": {
            "issues": "write",
            "pull_requests": "write",
            "contents": "write",
            "checks": "write",
            "metadata": "read",
            "members": "read",
        },
        "default_events": [
            "issues",
            "issue_comment",
            "pull_request",
            "pull_request_review",
            "pull_request_review_comment",
        ],
    }

    if _is_public_url(webhook_url):
        manifest["hook_attributes"] = {
            "url": f"{webhook_url}/github/webhooks",
            "active": True,
        }

    return manifest


# ---------------------------------------------------------------------------
# GitHub API
# ---------------------------------------------------------------------------

def exchange_code(code: str) -> dict:
    """Exchange a manifest creation code for app credentials."""
    resp = httpx.post(
        f"https://api.github.com/app-manifests/{code}/conversions",
        headers={"Accept": "application/vnd.github+json"},
    )
    if not resp.is_success:
        raise RuntimeError(f"GitHub API error: {resp.status_code} {resp.text}")
    return resp.json()


# ---------------------------------------------------------------------------
# Credential storage
# ---------------------------------------------------------------------------

def save_credentials(data: dict) -> None:
    """Save app credentials to ~/.config/clawcode/ and .env."""
    config_dir = Path.home() / ".config" / "clawcode"
    config_dir.mkdir(parents=True, exist_ok=True)

    pem_path = config_dir / "github-app.pem"
    pem_path.write_text(data["pem"])
    pem_path.chmod(0o600)
    logger.info("GitHub App private key saved", pem_path=str(pem_path))

    env_path = Path.cwd() / ".env"
    env_lines = (
        "\n"
        "# GitHub App (auto-configured)\n"
        f"GITHUB_APP_ID={data['id']}\n"
        f"GITHUB_WEBHOOK_SECRET={data['webhook_secret']}\n"
        f"GITHUB_PRIVATE_KEY_PATH={pem_path}\n"
    )
    with open(env_path, "a") as f:
        f.write(env_lines)

    logger.info(
        "GitHub App credentials saved to .env",
        app_id=data["id"],
        slug=data.get("slug"),
    )


# ---------------------------------------------------------------------------
# HTML pages
# ---------------------------------------------------------------------------

def _setup_page_html(webhook_url: str) -> str:
    manifest = build_app_manifest(webhook_url)
    manifest_json = json.dumps(manifest).replace("'", "&#39;")

    return f"""\
<!DOCTYPE html>
<html>
<head><title>ClawCode &mdash; Create GitHub App</title>
<style>
  body {{ font-family: system-ui; max-width: 600px; margin: 40px auto; padding: 0 20px; }}
  h1 {{ color: #333; }}
  .btn {{ background: #2ea44f; color: white; border: none; padding: 12px 24px;
         font-size: 16px; border-radius: 6px; cursor: pointer; }}
  .btn:hover {{ background: #2c974b; }}
  code {{ background: #f6f8fa; padding: 2px 6px; border-radius: 3px; }}
</style>
</head>
<body>
  <h1>ClawCode &mdash; Create GitHub App</h1>
  <p>Click below to create a GitHub App with the correct permissions.</p>
  <p>Webhook URL: <code>{webhook_url}/github/webhooks</code></p>
  <form action="https://github.com/settings/apps/new" method="post">
    <input type="hidden" name="manifest" value='{manifest_json}'>
    <button type="submit" class="btn">Create GitHub App</button>
  </form>
  <p><small>This will redirect you to GitHub to approve the app creation.</small></p>
</body>
</html>"""


def _success_page_html(slug: str, install_url: str) -> str:
    return f"""\
<!DOCTYPE html>
<html>
<head><title>ClawCode &mdash; Setup Complete</title>
<style>
  body {{ font-family: system-ui; max-width: 600px; margin: 40px auto; padding: 0 20px; }}
  h1 {{ color: #2ea44f; }}
  .btn {{ background: #2ea44f; color: white; border: none; padding: 12px 24px;
         font-size: 16px; border-radius: 6px; cursor: pointer; text-decoration: none;
         display: inline-block; }}
  code {{ background: #f6f8fa; padding: 2px 6px; border-radius: 3px; }}
</style>
</head>
<body>
  <h1>Setup Complete!</h1>
  <p>GitHub App <strong>{slug}</strong> has been created. Credentials saved to <code>.env</code>.</p>
  <p>Now install it on the repositories you want the bot to monitor:</p>
  <a href="{install_url}" class="btn">Install on Repositories</a>
  <p><small>You can close this tab after installing. Then restart ClawCode to load the new credentials.</small></p>
</body>
</html>"""


# ---------------------------------------------------------------------------
# Local HTTP server
# ---------------------------------------------------------------------------

class _SetupHandler(BaseHTTPRequestHandler):
    """Handles /setup and /callback for the one-shot manifest flow."""

    def __init__(self, *args, webhook_url: str, done_event: threading.Event, **kwargs):
        self.webhook_url = webhook_url
        self.done_event = done_event
        super().__init__(*args, **kwargs)

    def log_message(self, format, *args):  # noqa: A002
        # Silence default stderr logging from BaseHTTPRequestHandler
        pass

    def do_GET(self):  # noqa: N802
        parsed = urlparse(self.path)

        if parsed.path in ("/", "/setup"):
            self._serve_setup_page()
        elif parsed.path == "/callback":
            self._handle_callback(parsed.query)
        else:
            self.send_error(404)

    def _serve_setup_page(self):
        html = _setup_page_html(self.webhook_url)
        self.send_response(200)
        self.send_header("Content-Type", "text/html")
        self.end_headers()
        self.wfile.write(html.encode())

    def _handle_callback(self, query: str):
        params = parse_qs(query)
        code = params.get("code", [None])[0]
        if not code:
            self.send_response(400)
            self.send_header("Content-Type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Missing code parameter")
            return

        try:
            data = exchange_code(code)
            save_credentials(data)

            install_url = f"{data['html_url']}/installations/new"
            slug = data.get("slug", f"app-{data['id']}")

            html = _success_page_html(slug, install_url)
            self.send_response(200)
            self.send_header("Content-Type", "text/html")
            self.end_headers()
            self.wfile.write(html.encode())

            logger.info(
                "GitHub App created",
                app_id=data["id"],
                slug=slug,
                install_url=install_url,
            )
            self.done_event.set()

        except Exception as exc:
            logger.error("GitHub App manifest callback failed", error=str(exc))
            self.send_response(500)
            self.send_header("Content-Type", "text/plain")
            self.end_headers()
            self.wfile.write(f"Setup failed: {exc}".encode())
            self.done_event.set()


def run_setup_server(webhook_url: str) -> None:
    """Start a temporary local server, open the browser, and block until done."""
    done = threading.Event()
    handler = partial(_SetupHandler, webhook_url=webhook_url, done_event=done)
    server = HTTPServer(("127.0.0.1", SETUP_PORT), handler)

    setup_url = f"http://localhost:{SETUP_PORT}/setup"
    print(f"\nOpening browser to create GitHub App...\n  {setup_url}\n")
    webbrowser.open(setup_url)

    # Serve in a background thread so we can enforce the timeout
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()

    if not done.wait(timeout=TIMEOUT_SECONDS):
        logger.error("Setup timed out — no callback received within 5 minutes.")
        sys.exit(1)

    server.shutdown()


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(description="Create a GitHub App via the manifest flow")
    parser.add_argument(
        "--webhook-url",
        required=True,
        help="Public URL where GitHub will send webhooks (e.g. https://abc123.ngrok-free.app)",
    )
    args = parser.parse_args()

    # Already configured?
    env = read_env_file(["GITHUB_APP_ID"])
    if env.get("GITHUB_APP_ID"):
        logger.info(
            "GitHub App already configured",
            app_id=env["GITHUB_APP_ID"],
        )
        return

    # Validate URL
    parsed = urlparse(args.webhook_url)
    if not parsed.scheme or not parsed.hostname:
        logger.error("Invalid URL", url=args.webhook_url)
        sys.exit(1)

    run_setup_server(args.webhook_url)


if __name__ == "__main__":
    main()
